### ISSUE TITLE
Apollo Federation Example

### ISSUE THREAD

**Issue by jacob-lester** [Urigo/graphql-modules#673]
In the example federation project, I am getting an error when running this query through the gateway (i.e. `localhost:4000`) *(Note: This query does in fact work in the official [apollographql/federation-demo](https://github.com/apollographql/federation-demo) example)*

It's having an issue resolving the extended field `inStock` from the `inventory` service, although querying the inventory service directly (i.e. `localhost:4004`) correctly returns the data

```graphql
{
  topProducts {
    inStock
  }
}
```

```json
{
  "errors": [
    {
      "message": "Cannot read property 'session' of undefined",
      "path": [
        "_entities",
        0
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "serviceName": "inventory",
        "query": "query ($representations: [_Any!]!) {\n  _entities(representations: $representations) {\n    ... on Product {\n      inStock\n    }\n  }\n}",
        "variables": {
          "representations": [
            {
              "__typename": "Product",
              "upc": "1"
            },
            {
              "__typename": "Product",
              "upc": "2"
            },
            {
              "__typename": "Product",
              "upc": "3"
            }
          ]
        },
        "exception": {
          "stacktrace": [
            "TypeError: Cannot read property 'session' of undefined",
            "    at typeResolvers.(anonymous function) (C:\\Users\\lesterjd\\Documents\\apollo-federation\\node_modules\\@graphql-modules\\core\\src\\graphql-module.ts:1080:37)",
            "    at representations.map (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\@apollo\\federation\\src\\types.ts:99:22)",
            "    at Array.map (<anonymous>)",
            "    at resolve (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\@apollo\\federation\\src\\types.ts:82:28)",
            "    at field.resolve (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql-extensions\\src\\index.ts:274:18)",
            "    at resolveFieldValueOrError (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:503:18)",
            "    at resolveField (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:470:16)",
            "    at executeFields (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:311:18)",
            "    at executeOperation (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:255:122)",
            "    at executeImpl (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:102:14)"
          ]
        }
      }
    },
    {
      "message": "Cannot read property 'session' of undefined",
      "path": [
        "_entities",
        1
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "serviceName": "inventory",
        "query": "query ($representations: [_Any!]!) {\n  _entities(representations: $representations) {\n    ... on Product {\n      inStock\n    }\n  }\n}",
        "variables": {
          "representations": [
            {
              "__typename": "Product",
              "upc": "1"
            },
            {
              "__typename": "Product",
              "upc": "2"
            },
            {
              "__typename": "Product",
              "upc": "3"
            }
          ]
        },
        "exception": {
          "stacktrace": [
            "TypeError: Cannot read property 'session' of undefined",
            "    at typeResolvers.(anonymous function) (C:\\Users\\lesterjd\\Documents\\apollo-federation\\node_modules\\@graphql-modules\\core\\src\\graphql-module.ts:1080:37)",
            "    at representations.map (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\@apollo\\federation\\src\\types.ts:99:22)",
            "    at Array.map (<anonymous>)",
            "    at resolve (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\@apollo\\federation\\src\\types.ts:82:28)",
            "    at field.resolve (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql-extensions\\src\\index.ts:274:18)",
            "    at resolveFieldValueOrError (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:503:18)",
            "    at resolveField (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:470:16)",
            "    at executeFields (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:311:18)",
            "    at executeOperation (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:255:122)",
            "    at executeImpl (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:102:14)"
          ]
        }
      }
    },
    {
      "message": "Cannot read property 'session' of undefined",
      "path": [
        "_entities",
        2
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "serviceName": "inventory",
        "query": "query ($representations: [_Any!]!) {\n  _entities(representations: $representations) {\n    ... on Product {\n      inStock\n    }\n  }\n}",
        "variables": {
          "representations": [
            {
              "__typename": "Product",
              "upc": "1"
            },
            {
              "__typename": "Product",
              "upc": "2"
            },
            {
              "__typename": "Product",
              "upc": "3"
            }
          ]
        },
        "exception": {
          "stacktrace": [
            "TypeError: Cannot read property 'session' of undefined",
            "    at typeResolvers.(anonymous function) (C:\\Users\\lesterjd\\Documents\\apollo-federation\\node_modules\\@graphql-modules\\core\\src\\graphql-module.ts:1080:37)",
            "    at representations.map (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\@apollo\\federation\\src\\types.ts:99:22)",
            "    at Array.map (<anonymous>)",
            "    at resolve (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\@apollo\\federation\\src\\types.ts:82:28)",
            "    at field.resolve (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql-extensions\\src\\index.ts:274:18)",
            "    at resolveFieldValueOrError (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:503:18)",
            "    at resolveField (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:470:16)",
            "    at executeFields (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:311:18)",
            "    at executeOperation (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:255:122)",
            "    at executeImpl (C:\\Users\\lesterjd\\Documents\\apollo-federation\\services\\inventory\\node_modules\\graphql\\execution\\execute.js:102:14)"
          ]
        }
      }
    }
  ],
  "data": {
    "topProducts": [
      {
        "inStock": null
      },
      {
        "inStock": null
      },
      {
        "inStock": null
      }
    ]
  }
}
```

**Comment by ardatan** [comment]
Fixed the example; each server definition must have the following `context` declaration;
```ts
const server = new ApolloServer({
// ...
  context: session => session
})
```

**Comment by maapteh** [comment]
@jacob-lester did you try `context: session => ({ session }),`?

**Comment by jacob-lester** [comment]
Just tried it. Still same error

**Comment by ardatan** [comment]
@jacob-lester I updated the example and fixed the context issue so could you try with the following version?
`0.7.13-alpha-2a968e3.9+2a968e3`

**Comment by jacob-lester** [comment]
That appeared to work! Thanks for your quick response!

**Comment by jacob-lester** [comment]
Ok one more thing @ardatan When I try to implement federation in my own project, the `context.injector` is undefined when `__resolveReference` is called:

```ts
import { PersonProvider } from '../providers/person.provider';
import { Person } from 'types';

export default {
  Person: {
    __resolveReference(root: Person, _args: any, context: any, _info: any) {
      const provider: PersonProvider = context.injector.get(PersonProvider);
      return {
        ...root,
        ...provider.getPersonByUsername(root.username)
      }
    }
  }
}
```
Here is what the `context` object looks like:

```js
{
  fieldName: '_entities',
  fieldNodes: [{
    kind: 'Field',
    alias: undefined,
    name: [Object],
    arguments: [Array],
    directives: [],
    selectionSet: [Object],
    loc: [Object]
  }],
  returnType: [_Entity] !,
  parentType: Query,
  path: {
    prev: undefined,
    key: '_entities'
  },
  schema: GraphQLSchema {
    __validationErrors: [],
    __allowedLegacyNames: [],
    _queryType: Query,
    _mutationType: undefined,
    _subscriptionType: undefined,
    _directives: [@include,
      @skip,
      @deprecated,
      @key,
      @extends,
      @external,
      @requires,
      @provides
    ],
    astNode: undefined,
    extensionASTNodes: [],
    _typeMap: [Object: null prototype] {
      Query: Query,
      _Any: _Any,
      _Entity: _Entity,
      Person: Person,
      String: String,
      Int: Int,
      Submission: Submission,
      Deadline: Deadline,
      Boolean: Boolean,
      Opportunity: Opportunity,
      _Service: _Service,
      __Schema: __Schema,
      __Type: __Type,
      __TypeKind: __TypeKind,
      __Field: __Field,
      __InputValue: __InputValue,
      __EnumValue: __EnumValue,
      __Directive: __Directive,
      __DirectiveLocation: __DirectiveLocation
    },
    _possibleTypeMap: [Object: null prototype] {
      _Entity: [Object]
    },
    _implementations: [Object: null prototype] {
      _Entity: []
    },
    _extensionsEnabled: true
  },
  fragments: [Object: null prototype] {},
  rootValue: undefined,
  operation: {
    kind: 'OperationDefinition',
    operation: 'query',
    name: undefined,
    variableDefinitions: [
      [Object]
    ],
    directives: [],
    selectionSet: {
      kind: 'SelectionSet',
      selections: [Array],
      loc: [Object]
    },
    loc: {
      start: 0,
      end: 159
    }
  },
  variableValues: {
    representations: [
      [Object]
    ]
  },
  cacheControl: {
    setCacheHint: [Function: setCacheHint],
    cacheHint: {
      maxAge: 0
    }
  },
  session: {
    req: IncomingMessage {
      _readableState: [ReadableState],
      readable: false,
      _events: [Object],
      _eventsCount: 1,
      _maxListeners: undefined,
      socket: [Socket],
      connection: [Socket],
      httpVersionMajor: 1,
      httpVersionMinor: 1,
      httpVersion: '1.1',
      complete: true,
      headers: [Object],
      rawHeaders: [Array],
      trailers: {},
      rawTrailers: [],
      aborted: false,
      upgrade: false,
      url: '/graphql',
      method: 'POST',
      statusCode: null,
      statusMessage: null,
      client: [Socket],
      _consuming: true,
      _dumped: false,
      next: [Function: next],
      baseUrl: '',
      originalUrl: '/graphql',
      _parsedUrl: [Url],
      params: {},
      query: {},
      res: [ServerResponse],
      body: [Object],
      _body: true,
      length: undefined
    },
    res: ServerResponse {
      _events: [Object],
      _eventsCount: 2,
      _maxListeners: undefined,
      output: [],
      outputEncodings: [],
      outputCallbacks: [],
      outputSize: 0,
      writable: true,
      _last: false,
      chunkedEncoding: false,
      shouldKeepAlive: false,
      useChunkedEncodingByDefault: true,
      sendDate: true,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      _contentLength: null,
      _hasBody: true,
      _trailer: '',
      finished: false,
      _headerSent: false,
      socket: [Socket],
      connection: [Socket],
      _header: null,
      _onPendingData: [Function: bound updateOutgoingData],
      _sent100: false,
      _expect_continue: false,
      req: [IncomingMessage],
      locals: [Object: null prototype] {},
      _onceFinishListeners: [Array],
      [Symbol(isCorked)]: false,
      [Symbol(outHeadersKey)]: [Object]
    },
    _extensionStack: GraphQLExtensionStack {
      extensions: [Array]
    }
  }
}
```


However, when I just used a dummy array of users I searched in (instead of using injector/ person provider), I was able to get it working correctly:

```ts
import { PersonProvider } from '../providers/person.provider';
import { Person } from 'types';

const users = [
  {
    username: 'user1',
    firtname: 'Test',
    lastname: 'User'
  }
]

export default {
  Person: {
    __resolveReference(root: Person, _args: any, context: any, _info: any) {
      return {
        ...root,
        ...users.find(x => x.username == root.username)
      }
    }
  }
}
```

**Comment by jacob-lester** [comment]
The `context` object should look like this I believe:

```js
{
  injector: Injector {
    options: {
      name: 'PersonModule_SESSION',
      injectorScope: 'SESSION',
      hooks: [Array],
      children: []
    },
    _classMap: Map {
      [Function: PersonProvider] => [Function: PersonProvider]
    },
    _factoryMap: Map {},
    _applicationScopeInstanceMap: Map {
      Symbol(ModuleConfig.PersonModule) => undefined,
        [Function] => undefined,
        [Function: PersonProvider] => [PersonProvider]
    },
    _sessionScopeInstanceMap: Map {
      [Function: ModuleSessionInfo] => [ModuleSessionInfo]
    },
    _applicationScopeServiceIdentifiers: [Symbol(ModuleConfig.PersonModule),
      [Function],
      [Function: PersonProvider]
    ],
    _requestScopeServiceIdentifiers: [],
    _sessionScopeServiceIdentifiers: [
      [Function: ModuleSessionInfo]
    ],
    _hookServiceIdentifiersMap: Map {},
    _name: 'PersonModule_SESSION',
    _injectorScope: 'SESSION',
    _defaultProviderScope: 'SESSION',
    _hooks: ['onInit',
      'onRequest',
      'onResponse',
      'onError',
      'onConnect',
      'onOperation',
      'onOperationComplete',
      'onDisconnect'
    ],
    _initialProviders: [],
    _children: [],
    onInstanceCreated: [Function],
    _sessionSessionInjectorMap: WeakMap {
      [items unknown]
    }
  }
}
```

**Comment by ardatan** [comment]
@jacob-lester https://www.apollographql.com/docs/apollo-server/api/apollo-federation/
As you can see in Apollo docs, `__resolveReference` doesn't have `args` so its signature like below;
`__resolveReference(root, context, info)` not `__resolveReference(root, args, context, info`

**Comment by ardatan** [comment]
Available in 0.5.13!
